README.md
# ValidateNzBankAcc

An implementation of the process described on page 15 of
http://www.ird.govt.nz/resources/a/c/ac60890040b57d0e9bd4df41f9f3ce1d/rwt-nrwt-spec-2010-v3.pdf

## Installation

Add this line to your application's Gemfile:

    gem 'validate_nz_bank_acc'

And then execute:

    $ bundle install

Or install it yourself as:

    $ gem install validate_nz_bank_acc

## Usage

  ValidateNzBankAcc.new(bank_id, branch_id, account_number, suffix).valid?

## Contributing

1. Fork it
2. Create your feature branch (`git checkout -b my-new-feature`)
3. Commit your changes (`git commit -am 'Added some feature'`)
4. Push to the branch (`git push origin my-new-feature`)
5. Create new Pull Request


CHANGELOG
0.0.3 - Bugfix
0.0.2 - Never happened
0.0.1 - Initial Release


spec/spec_helper.rb
  require File.expand_path("../../lib/validate_nz_bank_acc.rb", __FILE__)


spec/validate_nz_bank_acc_spec.rb
require 'spec_helper'
describe ValidateNzBankAcc do

  let(:ex1) { ValidateNzBankAcc.new("01","902","-0068389",'00') }
  let(:ex2) { ValidateNzBankAcc.new("08","6523","1954512","001") }
  let(:ex3) { ValidateNzBankAcc.new("26","2600","0320871","032") }


  describe "valid_bank_code?" do
    it "should check against BANKS for the bank code" do
      ValidateNzBankAcc.new("03","0123","0034141","03").valid_bank_code?.should be_true
      ValidateNzBankAcc.new("01","1113","0034141","03").valid_bank_code?.should be_true
      ValidateNzBankAcc.new("20","0123","1111111","11").valid_bank_code?.should be_true
      ValidateNzBankAcc.new("05","0123","0034141","03").valid_bank_code?.should be_false # no back 05
    end
  end

  describe "valid_bank_branch?" do
    it "should validate the branch code against the range of valid branches" do
      # VALIDS
      ValidateNzBankAcc.new("03","0123","0034141","03").valid_branch_code?.should be_true
      ValidateNzBankAcc.new("26","2600","0034141","03").valid_branch_code?.should be_true
      ValidateNzBankAcc.new("26","2699","0034141","0003").valid_branch_code?.should be_true
      ValidateNzBankAcc.new("11","6666","0034141","0003").valid_branch_code?.should be_true

      # INVALIDS
      ValidateNzBankAcc.new("11","1111","0034141","0003").valid_branch_code?.should be_false
      ValidateNzBankAcc.new("01","2012","0034141","0003").valid_branch_code?.should be_false
    end
  end

  describe "algo_code" do
    it "should select the correct algo_code code" do
      ValidateNzBankAcc.new("08","0123","0034141","03").algo_code.should == :d
      ValidateNzBankAcc.new("31","0123","0034141","03").algo_code.should == :x

      # If the account base number is below 00990000 then apply algorithm A, otherwise apply algorithm B
      ValidateNzBankAcc.new("30","0123","0034141","03").algo_code.should == :a
      ValidateNzBankAcc.new("30","0123","1034141","03").algo_code.should == :b
      ValidateNzBankAcc.new("30","0123","0990000","03").algo_code.should == :b
      ValidateNzBankAcc.new("30","0123","0989999","03").algo_code.should == :a
    end
    it "should select the correct algo_code code according to the pdf" do
      ex1.algo_code.should == :a
      ex2.algo_code.should == :d
      ex3.algo_code.should == :g
    end
  end

  describe "number_18_char" do
    it "should zero pad the 7 digit account number to 8 characters" do
      ValidateNzBankAcc.new("08","0123","0034141","03").number.length.should == 18
      ValidateNzBankAcc.new("08","0123","0034141","03").number.should == '080123000341410003'
    end
  end

  describe "checksum_sum" do
    it "should return the same numbers as in the pdf" do
      ex1.checksum_sum.should == 176
      ex2.checksum_sum.should == 121
      ex3.checksum_sum.should == 30
    end
  end

  describe "valid_modulo?" do
    it "should valid modulo for the example in the pdf" do
      ex1.valid_modulo?.should be_true
      ex1.valid_modulo?.should be_true
      ex1.valid_modulo?.should be_true
    end
  end
end

lib/validate_nz_bank_acc.rb
require "validate_nz_bank_acc/version"

class ValidateNzBankAcc

  BANKS = { 1 => {:ranges => [1..999, 1100..1199, 1800..1899]},
            2 => {:ranges => [1..999, 1200..1299]},
            3 => {:ranges => [1..999, 1300..1399, 1500..1599, 1700..1799, 1900..1999]},
            6 => {:ranges => [1..999, 1400..1499]},
            8 => {:ranges => [6500..6599], :algo => :d},
            9 => {:ranges => [], :algo => :e}, # range was 0000
            11 => {:ranges => [5000..6499, 6600..8999]},
            12 => {:ranges => [3000..3299, 3400..3499,3600..3699]},
            13 => {:ranges => [4900..4999]},
            14 => {:ranges => [4700..4799]},
            15 => {:ranges => [3900..3999]},
            16 => {:ranges => [4400..4499]},
            17 => {:ranges => [3300..3399]},
            18 => {:ranges => [3500..3599]},
            19 => {:ranges => [4600..4649]},
            20 => {:ranges => [4100..4199]},
            22 => {:ranges => [4000..4049]},
            23 => {:ranges => [3700..3799]},
            24 => {:ranges => [4300..4349]},
            25 => {:ranges => [2500..2599], :algo => :f},
            26 => {:ranges => [2600..2699], :algo => :g},
            27 => {:ranges => [3800..3849]},
            28 => {:ranges => [2100..2149], :algo => :g},
            29 => {:ranges => [2150..2299], :algo => :g},
            30 => {:ranges => [2900..2949]},
            31 => {:ranges => [2800..2849], :algo => :x},
            33 => {:ranges => [6700..6799], :algo => :f},
            35 => {:ranges => [2400..2499]},
            38 => {:ranges => [9000..9499]}
          }

  ALGOS = {
           :a =>[0, 0, 6, 3, 7, 9, 0, 0, 10, 5, 8, 4, 2, 1, 0, 0, 0, 0, 11],
           :b => [0, 0, 0, 0, 0, 0, 0, 0, 10, 5, 8, 4, 2, 1, 0, 0, 0, 0, 11],
           :c => [3, 7, 0, 0, 0, 0, 9, 1, 10, 5, 3, 4, 2, 1, 0, 0, 0, 0, 11],
           :d => [0, 0, 0, 0, 0, 0, 0, 7, 6, 5, 4, 3, 2, 1, 0, 0, 0, 0, 11],
           :e => [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 5, 4, 3, 2, 0, 0, 0, 1, 11],
           :f => [ 0, 0, 0, 0, 0, 0, 0, 1, 7, 3, 1, 7, 3, 1, 0, 0, 0, 0, 10],
           :g => [0, 0, 0, 0, 0, 0, 0, 1, 3, 7, 1, 3, 7, 1, 0, 3, 7, 1, 10],
           :x => [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1]
          }
  attr_accessor :bank_id, :branch_id, :account_number, :suffix
  def initialize(bank_id, branch_id, account_number, suffix)
    @bank_id = process_input(bank_id)
    @branch_id = process_input(branch_id)
    @account_number = process_input(account_number)
    @suffix = process_input(suffix)
  end

  def valid?
    valid_branch_code? && valid_modulo?
  end

  def valid_bank_code?
    BANKS.has_key? bank_id
  end

  def valid_branch_code?
    return false unless valid_bank_code?
    if BANKS[bank_id][:ranges].empty?
      # Bank 9. Anything is valid? # DOUBLE CHECK
      true
    else
      BANKS[bank_id][:ranges].any? do |range|
        range.include? branch_id
      end
    end
  end

  def algo_code
    # If the account base number is below 00990000 then apply algorithm A, otherwise apply algorithm B.
    BANKS[bank_id][:algo] || (account_number < 990000 ? :a : :b)
  end

  def algo
    ALGOS[algo_code]
  end

  def checksum_sum
    if [:e, :g].include? algo_code
      (0..17).inject(0) do |sum, index|
        s = number[index].to_i * algo[index]
        2.times { s = s.to_s.chars.map(&:to_i).inject(:+) }
        sum += s
      end
    else
      (0..17).inject(0) {|sum, index| sum += number[index].to_i * algo[index]; sum }
    end
  end

  def valid_modulo?
    checksum_sum % algo[18] == 0
  end

  def number # account number padded to 8 chars
    bank_code + branch_code + account_number_code + suffix_code
  end

  private

    def bank_code
      @bank_id.to_s.rjust(2,'0')
    end

    def branch_code
      @branch_id.to_s.rjust(4,'0')
    end

    def account_number_code
      @account_number.to_s.rjust(8,'0')
    end

    def suffix_code
      @suffix.to_s.rjust(4,'0')
    end

    def process_input(num)
      num.to_s.gsub(/[^0-9]/,'').to_i
    end
end
